### PR TITLE
Document parts of Rust runtime [ECR-3285]

### PR DESCRIPTION
**Pull request by alekseysidorov**

### CODE REVIEW

File: exonum/src/runtime/rust/mod.rs [hunk 1]
```diff
@@ -54,6 +67,9 @@ mod service;
 #[cfg(test)]
 mod tests;
 
+/// Rust runtime instance.
+///
+/// [Read more about the Rust runtime](index.html).
```
Comment (popzxc) [inline exonum/src/runtime/rust/mod.rs]:
```suggestion
/// [Detailed description of the Rust runtime](index.html).
```

File: exonum/src/runtime/rust/mod.rs [hunk 2]
```diff
@@ -54,6 +67,9 @@ mod service;
 #[cfg(test)]
 mod tests;
 
+/// Rust runtime instance.
```
Comment (popzxc) [inline exonum/src/runtime/rust/mod.rs]:
```suggestion
/// Rust runtime entity.
```

File: exonum/src/runtime/rust/mod.rs [hunk 3]
```diff
@@ -12,7 +12,20 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
-//! Built-in Rust runtime module.
+//! The basic runtime with the native Rust services.
```
Comment (popzxc) [inline exonum/src/runtime/rust/mod.rs]:
```suggestion
//! The runtime for the native Rust services.
```

File: exonum/src/runtime/rust/mod.rs [hunk 4]
```diff
@@ -12,7 +12,20 @@
...
+//! During this procedure, be very careful and avoid situations when some of deployed services do
+//! not exist in the list of available. The easiest way to add only a new available artifacts, rather
+//! than delete.
+//!
+//! Rust runtime does not provide any level of service isolation from the operation system, therefore
+//! you should only use a trusted artifacts.
+//!
+//! Artifacts available for deployment are presented by the [`ServiceFactory`][ServiceFactory] objects.
+//!
+//! [ServiceFactory]: trait.ServiceFactory.html
```
Comment (popzxc) [inline exonum/src/runtime/rust/mod.rs]: Not required anymore with [intra-rustdoc links](https://github.com/rust-lang/rust/issues/43466).

File: exonum/src/runtime/rust/mod.rs [hunk 5]
```diff
@@ -12,7 +12,20 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
-//! Built-in Rust runtime module.
+//! The basic runtime with the native Rust services.
+//!
+//! This runtime is usually presents in every blockchain instance and can only process a static set
```
Comment (popzxc) [inline exonum/src/runtime/rust/mod.rs]: >usually presents in every blockchain instance

Usually or every? And it's not clear what is "blockchain instance" (`Blockchain` object, the node, the network).

File: exonum/src/runtime/rust/mod.rs [hunk 6]
```diff
@@ -12,7 +12,20 @@
...
+//! of available Rust services. To change this set you have to recompile the node binary.
```
Comment (popzxc) [inline exonum/src/runtime/rust/mod.rs]:
```suggestion
//! of available Rust services. This set is defined at the compile-time level. Once created, it can be changed only by the node binary recompilation.
```

File: exonum/src/runtime/rust/mod.rs [hunk 7]
```diff
@@ -12,7 +12,20 @@
...
+//! During this procedure, be very careful and avoid situations when some of deployed services do
```
Comment (popzxc) [inline exonum/src/runtime/rust/mod.rs]: I'd rephrase like "Beware of the removing the artifacts from the Rust runtime, since attempt to remove an artifact with already running instances can cause blockchain to break. The only safe change that can be performed here is adding new artifacts."

File: exonum/src/runtime/rust/mod.rs [hunk 8]
```diff
@@ -12,7 +12,20 @@
...
+//! not exist in the list of available. The easiest way to add only a new available artifacts, rather
+//! than delete.
+//!
+//! Rust runtime does not provide any level of service isolation from the operation system, therefore
+//! you should only use a trusted artifacts.
```
Comment (popzxc) [inline exonum/src/runtime/rust/mod.rs]:
```suggestion
//! security audit of the artifacts to be deployed is up to the node administrators.
```

File: exonum/src/runtime/rust/mod.rs [hunk 9]
```diff
@@ -12,7 +12,20 @@
...
+//!
+//! Artifacts available for deployment are presented by the [`ServiceFactory`][ServiceFactory] objects.
```
Comment (popzxc) [inline exonum/src/runtime/rust/mod.rs]: I'd expand this sentence with some explanation of what an "artifact available to deploy" is, and how `ServiceFactory` helps to achieve it.

File: exonum/src/runtime/rust/mod.rs [hunk 10]
```diff
@@ -121,12 +137,15 @@ impl RustRuntime {
             .expect("Method called before Rust runtime is initialized")
     }
 
+    /// Adds a new service factory to the list of available artifacts.
```
Comment (popzxc) [inline exonum/src/runtime/rust/mod.rs]: This doc comment can lead to a conclusion that "service factory" is an "artifact", which is a bit confusing.

File: exonum/src/runtime/rust/mod.rs [hunk 11]
```diff
@@ -121,12 +137,15 @@ impl RustRuntime {
...
     pub fn add_service_factory(&mut self, service_factory: Box<dyn ServiceFactory>) {
         let artifact = service_factory.artifact_id();
         trace!("Added available artifact {}", artifact);
         self.available_artifacts.insert(artifact, service_factory);
     }
 
+    /// Adds a new service factory to the list of available artifacts and returns
```
Comment (popzxc) [inline exonum/src/runtime/rust/mod.rs]: Same as above + I'd change "returns itself" to "returns modified `RustRuntime` object".

File: exonum/src/runtime/rust/mod.rs [hunk 12]
```diff
@@ -221,14 +240,27 @@ impl From<RustRuntime> for (u32, Box<dyn Runtime>) {
     }
 }
 
+/// The unique name of the Rust artifact, contains the name and version of the artifact.
```
Comment (popzxc) [inline exonum/src/runtime/rust/mod.rs]:
```suggestion
/// The unique identifier of the Rust artifact, containing the name and version of the artifact.
```

File: exonum/src/runtime/rust/mod.rs [hunk 13]
```diff
@@ -221,14 +240,27 @@ impl From<RustRuntime> for (u32, Box<dyn Runtime>) {
...
+///
+/// In string representation the artifact name is written as follows:
```
Comment (popzxc) [inline exonum/src/runtime/rust/mod.rs]:
```suggestion
/// As a string, the artifact name is represented as follows:
```

File: exonum/src/runtime/rust/mod.rs [hunk 14]
```diff
@@ -221,14 +240,27 @@ impl From<RustRuntime> for (u32, Box<dyn Runtime>) {
...
+///
+/// `{artifact_name}:{artifact_version}`, where `artifact_name` is a unique name of the artifact,
+/// and 'artifact_version` is a version number.
```
Comment (popzxc) [inline exonum/src/runtime/rust/mod.rs]:
```suggestion
/// and `artifact_version` is a semantic version identifier.
```

File: exonum/src/runtime/rust/mod.rs [hunk 15]
```diff
@@ -221,14 +240,27 @@ impl From<RustRuntime> for (u32, Box<dyn Runtime>) {
...
+///
+/// * Artifact name contains only the following characters: `a-zA-Z0-9` and one of `_-.`.
```
Comment (popzxc) [inline exonum/src/runtime/rust/mod.rs]:
```suggestion
/// * Artifact name can contain only the following characters: `a-zA-Z0-9` and one of `_-.`.
```

File: exonum/src/runtime/rust/mod.rs [hunk 16]
```diff
@@ -221,14 +240,27 @@ impl From<RustRuntime> for (u32, Box<dyn Runtime>) {
...
+/// * Artifact version number must conform to the semantic version scheme.
```
Comment (popzxc) [inline exonum/src/runtime/rust/mod.rs]:
```suggestion
/// * Artifact version identifier must conform to the semantic version scheme (major.minor.patch).
```

File: exonum/src/runtime/rust/mod.rs [hunk 17]
```diff
@@ -221,14 +240,27 @@ impl From<RustRuntime> for (u32, Box<dyn Runtime>) {
...
 #[derive(Debug, Clone, PartialEq, Eq, Hash)]
 pub struct RustArtifactId {
+    /// Artifact name.
     pub name: String,
+    /// Artifact version number conforming to the semantic versioning scheme.
```
Comment (popzxc) [inline exonum/src/runtime/rust/mod.rs]:
```suggestion
    /// Artifact version identifier conforming to the semantic versioning scheme.
```

File: exonum/src/runtime/rust/mod.rs [hunk 18]
```diff
@@ -221,14 +240,27 @@ impl From<RustRuntime> for (u32, Box<dyn Runtime>) {
...
     pub version: Version,
 }
 
 impl RustArtifactId {
+    /// Creates a new Rust artifact id from the specified parts.
```
Comment (popzxc) [inline exonum/src/runtime/rust/mod.rs]:
```suggestion
    /// Creates a new Rust artifact ID from the provided name and version.
```

File: exonum/src/runtime/rust/service.rs [hunk 19]
```diff
@@ -235,6 +235,7 @@ impl<'a> AfterCommitContext<'a> {
     }
 }
 
+/// Special extensions for the service instance authorized as a supervisor.
```
Comment (popzxc) [inline exonum/src/runtime/rust/service.rs]:
```suggestion
/// Extended blockchain interface for the service instance authorized as a supervisor.
```

File: exonum/src/runtime/rust/mod.rs [hunk 20]
```diff
@@ -12,7 +12,23 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
-//! Built-in Rust runtime module.
+//! The runtime for the native Rust services.
```
Comment (aleksuss) [inline exonum/src/runtime/rust/mod.rs]:
```suggestion
//! The current runtime is for running native services written in Rust.
```

File: exonum/src/runtime/rust/service.rs [hunk 21]
```diff
@@ -235,6 +235,7 @@ impl<'a> AfterCommitContext<'a> {
     }
 }
 
+/// Extended blockchain interface for the service instance authorized as a supervisor.
```
Comment (aleksuss) [inline exonum/src/runtime/rust/service.rs]: Extends ?

Comment (popzxc) [reply exonum/src/runtime/rust/service.rs]: It's a struct, so I guess `Extended` here is OK.

Comment (aleksuss) [reply exonum/src/runtime/rust/service.rs]: In depends on what `SupervisorExtensions` exactly is. If it's already extended interface than it's ok, but if it extends interface by itself than `Extends` is more appropriate.

Comment (popzxc) [reply exonum/src/runtime/rust/service.rs]: Well, I still find confusing a verb comment for a struct. Maybe then use "Extensions to the blockchain interface intended for the service instance authorized as a supervisor"?

File: exonum/src/runtime/rust/mod.rs (hunk 20 above)
Comment (popzxc) [reply exonum/src/runtime/rust/mod.rs]: I'm not sure about this. Why is the original line bad?

File: exonum/src/runtime/rust/service.rs (hunk 21 above)
Comment (aleksuss) [reply exonum/src/runtime/rust/service.rs]: Sounds good as for me. Maybe @Maria-Nosyk can help with it.

File: exonum/src/runtime/rust/mod.rs (hunk 20 above)
Comment (aleksuss) [reply exonum/src/runtime/rust/mod.rs]: Because I'm not sure that original line is grammatically correct 😄 @Maria-Nosyk help with it.
